### ISSUE TITLE
FormData send does not work

### ISSUE THREAD

**Issue by velochy** [oracle/cordova-plugin-wkwebview-file-xhr#4]
the error message states that body.keys() is not a function. Looking at FormData prototype in safari reveals that the only function FormData has implemented is append and keys() is indeed missing.

Thankfully, I found a polyfill that adds it at https://github.com/jimmywarting/FormData

For completeness, you may want to include this in your project though, or at least link to it in the readme.

**Comment by gvanmat** [comment]
Good point.  Thanks for the feedback.

**Comment by velochy** [comment]
The polyfill I linked to has problems when used with this plugin. After trying different combinations, the working solution involved these two steps:

a) once the form is ready, call fd._blob() to get a Blob and pass that as xhr request data
b) take the content type of that blob and set that as the contentType header 

Im guessing setting the contentType header to undefined will make it default to the content type set for the blob but I did not test that. If you do decide to integrate that polyfill, do pay attention to this though ;)

**Comment by velochy** [comment]
Also, on iOS the polyfill _.blob function is currently broken but this is easy to hotfix around by always using the placeholder "File" definition he provides and renaming it to BlobWrapper in his implementation (https://github.com/jimmywarting/FormData/issues/26)

**Comment by gvanmat** [comment]
@velochy Just checked in a FormData polyfill.  This is our own impl that works with the xhr polyfill. Please help me verify the fix.

**Comment by velochy** [comment]
@gvanmat Thank you! I actually had a working workaround but the code was ugly to say the least so if this works, I am really grateful! I will try to find time to check it over the coming week.

**Comment by gvanmat** [comment]
@velochy Just pushed in a few fixes to the FormData polyfill.  Missed the forEach method and had a couple more typos.

**Comment by velochy** [comment]
@gvanmat Sadly, it is still not posting the data properly.

I'm using AngularJS $http (wrapper for XmlHttpRequest) to post (roughly) the following:
formData = new FormData();
formData.append('nonce','xyz');
formData.append('file',Blob(['xyz']),'file.bin')
requestParams = {
          method: 'POST',
          url: url,
          data: formData,
          headers: { 'Content-Type': undefined},
          withCredentials: true
        };

What my server endpoint (Flask in Python) sees is:

HEADERS: [('Cookie', u'session=.eJw1j8FqwzAQRH-l7LmHRtQXQw8uMsaHtbBRK3YvoU3UOLIVipNgSSH_XrXQ2zDwmDc32H4t9jxCeVmu9hG2xz2UN3j4hBJYtwJNH8gNHtMhonydMc2O9XBEN8zc1JFMu5LvXGdQsGFPjp6UxkLJ_YiJJ3TVMzuMnaOk9PtIqY3o69g1tcjcRGIYc07odoH1VLDuM3_YsMmd6NfMBpa0od9NURdK70T2WlXzFrjpBcoqKFm9wD27f9vFf5zs6fL_5nq2y98jKOD-A3VATo0.DOYF8Q.-ceAB5JZx18GeGdF3EcbFqJ4nFU; remember_token=WyI1IiwiZDQxZDhjZDk4ZjAwYjIwNGU5ODAwOTk4ZWNmODQyN2UiXQ.DOXrOA.hoto_vO7QbdS7KrrQ1g5AbKlq9E'), ('Content-Length', u'11'), ('User-Agent', u'Achieve%20Music/1.1.6 CFNetwork/887 Darwin/17.0.0'), ('X-Forwarded-Proto', u'https'), ('Host', u'mms.achievemusic.com'), ('Accept', u'application/json, text/plain, */*'), ('Accept-Language', u'en-us'), ('Content-Type', u'text/plain;'), ('Accept-Encoding', u'br, gzip, deflate')]
DATA: '{"_map":{}}'

In short, it seems it is not really converting it to the multipart format as it should. 
Let me know if I can help with any more debug information.

**Comment by gvanmat** [comment]
The FormData is going to send multipart/form data.  The content-type of the request will contain a boundary token the back-end uses to parse the document parts.  So, you will want to remove the content-type header on the request.

However, it looks to me like angularjs $http service is calling JSON.stringify on the formData object versus passing it on the the XMLHttpRequest.send(formData).

What if you tried using the native XMLHttpRequest here just to help isolate the fault. Something like this:

```
formData = new FormData();
formData.append('nonce','xyz');
formData.append('file',Blob(['xyz']),'file.bin')
var myhttp = new Promise(function (resolve, reject)
{
  var xhr = new XMLHttpRequest();
  xhr.onerror = function (error) {reject(xhr.responseText)});
  xhr.onloadend = function (error) {reject(xhr.responseText)});
  xhr.open("POST", url);
  xhr.send(formData);
})

myhttp.then(function (text) 
{
)).catch(function (error)
{
});

```
Or, it could be the call to identify the body/object passed to the send method is failing to see it as a FormData prototype - FormData.prototype.isPrototypeOf(body).

At any rate, removing the request content-type header will help isolate.  If the FormData prototype check is failing, the server should see a request content-type of "application/json".  It plugins in default request header content-types if not provided.

**Comment by velochy** [comment]
@gvanmat it seems you are right regarding serialization. https://docs.angularjs.org/api/ng/service/$http "Default transformations" describes exactly that behavior.

Its easy to disable though. will test this tomorrow.

Considering the popularity of angularjs, you may want to make a note about it in the readme if it does work, though

**Comment by gvanmat** [comment]
@velochy I hope that's it.  Please let me know.  I suspect angular would have the same issue if you passed in a native javascript FormData.  For example, from a safari console: 

var fd = new FormData(); fd.append("foo", "bar"); JSON.stringify(fd);

Would result in "{}".

Since the Blob and File types need the FileReader to convert into a binary string, the conversion is async.  I would have liked to return the multipart body via toString() but needs a promise for the async conversion.

**Comment by velochy** [comment]
@gvanmat it got us a step forward but there is a new problem. The post is now properly parsed on the server, but the binary file being transmitted comes out broken. 

The file going into the FormData object in JS is a binary blob that, in base64 is:
AAAABSoAAAAAAAABEwAAAAcAAAAuAAAAhQAAAAIAAALwAAAAmgAAACYBAAEAAQABAAEAAAABAAEAAQABAAEAAQABAAEAAQABAAEAAQABAgAAACeur6+zs76+wMDNzbm5mJh+fn19dnZzc3Fxbm5zc3V1dnZ2dnd3dXRAuIyaAAAAmggYJEWQ3v/gjDEA/BdjvNaaPggBHGS82aFFDgYZVa7Xp0wQAxNNqdirTxAEDBIRDw8RExQYHR4hNFVsbmRbW2NnXEYxKSclKTEyJRgXHSEiKjAuKSkyPkZOVFFDMiorLiwnIB4nND5LVldPRkE2IxQRGR8jLjg5MCMeISg0QEVDQkRFSUo9Jx0hJSYqLTM3NjMvLi0lGxYYGRgAAAFAKgAAAAEAAAALAAAB/////6WHdcs=

What the server receives is a binary file that is somewhat changed (again, base64):
AAAABSoAAAAAAAABEwAAAAcAAAAuAAAAwoUAAAACAAACw7AAAADCmgAAACYBAAEAAQABAAEAAAABAAEAAQABAAEAAQABAAEAAQABAAEAAQABAgAAACfCrsKvwq/Cs8Kzwr7CvsOAw4DDjcONwrnCucKYwph+fn19dnZzc3Fxbm5zc3V1dnZ2dnd3dXRAwrjCjMKaAAAAwpoIGCRFwpDDnsO/w6DCjDEAw7wXY8K8w5bCmj4IARxkwrzDmcKhRQ4GGVXCrsOXwqdMEAMTTcKpw5jCq08QBAwSEQ8PERMUGB0eITRVbG5kW1tjZ1xGMSknJSkxMiUYFx0hIiowLikpMj5GTlRRQzIqKy4sJyAeJzQ+S1ZXT0ZBNiMUERkfIy44OTAjHiEoNEBFQ0JERUlKPScdISUmKi0zNzYzLy4tJRsWGBkYAAABQCoAAAABAAAACwAAAcO/w7/Dv8O/wqXCh3XDiw==

Since the multipart encoding works now, and other FormData parameters are properly passed, I'm inclined to think this is not related to angular. My best bet would be that the string gets scrambled being sent from javascript to native code.

Regarding AngularJS - it does not have the same issue with native FormData, presumably because it typechecks it somehow. So you may still want to add a line to the readme saying that for AngularJS (and potentially other frameworks) you may need to disable default transformations (for AngularJS, it means passing an extra option transformRequest: function(x) {return x;} to $http calls)

**Comment by gvanmat** [comment]
When you create your blob, what context-type are you assigning to the blob?  I'll do some more investigating with different file types.  I'm using the Java EE multipart form servlet in my testing so I know the server side handling is correct and should catch any issues.

If angular is checking the prototype of the data object, it should also work the same since we have replaced the window.FormData with our own instance.

**Comment by gvanmat** [comment]
@velochy I'm a fan of angularJS too.  I like the simplicity of creating components and the separation of  controllers with inject-able services.

I was looking in the $http service and you are correct about special [handling of the FormData](https://github.com/angular/angular.js/blob/master/src/ng/http.js#L325).  I found the [isFormData ](https://github.com/angular/angular.js/blob/master/src/Angular.js#L713) function and see what I'm doing wrong. 

The FormData polyfill toString returns "[object, FormData]" but this code is looking for "[object FormData]".  No comma.  I can fix that up so there is no need for any special code.

**Comment by velochy** [comment]
@gvanmat regarding blob content type - i do not set it explicitly and it seems to default to binary/octet-stream.
The blob is created pretty directly from base64 data in my own code so you should be able to reproduce with

FD.append ('file',new Blob([btoa ('<copy the above data>')]),'file.bin')

Hope it helps.

**Comment by velochy** [comment]
@gvanmat any news on the blobs getting mangled issue? Is there any other information I could provide?

**Comment by gvanmat** [comment]
@velochy I wasn't able to recreate the issue.  I created a [test case](https://github.com/oracle/cordova-plugin-wkwebview-file-xhr/blob/master/tests/tests.js#L169).  I'm uploading several blobs in the FormData.  The one that is base64 encoded returns the expected result.

What other variables do you have in your usecase outside of the angular http service?

**Comment by velochy** [comment]
@gvanmat thank you for testing it out. I will ask one of my colleagues for another pair of eyes to see if we can figure out what might be the reason. The fact is that the same code works in desktop browsers and android webviews so it has to be something ios related, but it may well be somewhere in our code. Will keep you posted.

**Comment by velochy** [comment]
@gvanmat just to make sure: did you test with the same b64 encoded string i sent? The test case you link has a simple ascii string which might not reveal encoding-related issues.

**Comment by gvanmat** [comment]
@velochy  I really appreciate your help verifying the FormData polyfill.  I'm pretty sure I have it this time.  Please try the latest again.

**Comment by velochy** [comment]
I can confirm it does indeed. Thank you for fixing it, @gvanmat. Great plugin and keep up the good work.

**Comment by Ralpharoo** [comment]
Note to any others who land here, this plugin breaks my signed AWS S3 uploads with blobs in FormData. I receive a request denied error from Amazon where something from the FormData signed request / file is lost in transmission. Have not spent much time digging though.

After removing this again, everything works perfectly (and vice versa). If it is of interest or affects others, I can try and set up a test case. Relevant or not, I am using $.ajax to upload the files and my own FormData object (with processData: false).

**Comment by gvanmat** [comment]
The plugin has a polyfill for [FormData](https://github.com/oracle/cordova-plugin-wkwebview-file-xhr/blob/master/src/www/ios/formdata-polyfill.js).  Sending the form data will result in the request being a multipart document.  I suspect AWS expects a multipart form.

Could you try using the XMLHttpRequest instead of the jquery wrapper function? It would be good to understand if it's an issue with the jquery integration or the plugin code.  Using the XMLHttpRequest isn't that much more verbose.

In addition, you might also take a couple seconds and enable plugin logging.  The [NativeXHRLogging preference](https://github.com/oracle/cordova-plugin-wkwebview-file-xhr#configuration) with a value of "full" will give you a bunch of useful information written to the JS console.

**Comment by Ralpharoo** [comment]
Thanks @gvanmat - Quick test shows this to be working fine using the XMLHttpRequest. For some reason I thought I tried this before but must not have!

**Comment by gvanmat** [comment]
I suspect the jQuery code looks at the prototype of the FormData or a toString check to identify it over a plain json object and the Pollyfill doesn't pass the check. Thanks for verifying the request object works.
